# Notes for the Haskell course (Funktionale Programmierung)

## Types and Typeclasses
- The type of every expressions is known at compile time.
- **Types** are capital! e.g: Integer
- **Typevariables** are lower case:

In [19]:
:t head

head :: forall a. [a] -> a

- Here [a] is a **typevariable** in a list.
- **Polymorphic functions**: Are functions that have **typevariables**.
- **Typeclass**: Interface that defines some behaviour:

In [20]:
:t (==)

(==) :: forall a. Eq a => a -> a -> Bool

- **Class Constraint**: _Eq a_ 
- Read it as following: The equality function takes any two values that are of the same type and returns a _Bool_. The type of those two values must be a member of the _Eq_ class.

## Syntax in functions
### Pattern matching

In [21]:
lucky :: (Integral a) => a -> String  
lucky 7 = "LUCKY NUMBER SEVEN!"  
lucky x = "Sorry, you're out of luck, pal!"

In [22]:
lucky 2

"Sorry, you're out of luck, pal!"

In [23]:
lucky 7

"LUCKY NUMBER SEVEN!"

When you call _lucky_, the patterns will be checked from top to bottom and when it conforms to a pattern, the corresponding function body will be used.

Another example:
We defined the factorial of a number n as product [1..n]. We can also define a factorial function recursively, the way it is usually defined in mathematics. We start by saying that the factorial of 0 is 1. Then we state that the factorial of any positive integer is that integer multiplied by the factorial of its predecessor. Here's how that looks like translated in Haskell terms.

In [24]:
factorial :: (Integral a) => a -> a  
factorial 0 = 1  
factorial n = n * factorial (n - 1) 

In [25]:
factorial 3

6

Explanation: If we try to get the _factorial_ of, say, 3. It tries to compute `3 * factorial 2`. The _factorial_ of 2 is `2 * factorial 1`, so for now we have `3 * (2 * factorial 1)`. _factorial_ 1 is `1 * factorial 0`, so we have `3 * (2 * (1 * factorial 0))`. Now here comes the trick — we've defined the _factorial_ of 0 to be just 1 and because it encounters that pattern before the catch-all one, it just returns 1. So the final result is equivalent to `3 * (2 * (1 * 1))`.

- Pattern matching and tuples:

In [26]:
addVectors :: (Num a) => (a, a) -> (a, a) -> (a, a)  
addVectors (x1, y1) (x2, y2) = (x1 + x2, y1 + y2)

In [27]:
addVectors (1, 1) (2, 2)

(3,3)

Another example

In [28]:
first :: (a, b, c) -> a  
first (x, _, _) = x  
  
second :: (a, b, c) -> b  
second (_, y, _) = y  
  
third :: (a, b, c) -> c  
third (_, _, z) = z  

The `_` means the same thing as it does in list comprehensions. It means that we really don't care what that part is, so we just write a `_`.

### Guards
Whereas patterns are a way of making sure a value conforms to some form and
deconstructing it, guards are a way of testing whether some property of a value
(or several of them) are true or false.

Note: Because iHaskell only takes single line input, the bmiTell function is written out here and underneath this to try it out.
```Haskell
bmiTell :: (RealFloat a) => a -> String
bmiTell bmi
| bmi <= 18.5 = "You're underweight, you emo, you!"
| bmi <= 25.0 = "You're supposedly normal. Pffft, I bet you're ugly!"
| bmi <= 30.0 = "You're fat! Lose some weight, fatty!"
| otherwise   = "You're a whale, congratulations!"
```

In [29]:
bmiTell :: (RealFloat a) => a -> String
bmiTell bmi | bmi <= 18.5 = "You're underweight, you emo, you!" | bmi <= 25.0 = "You're supposedly normal. Pffft, I bet you're ugly!" | bmi <= 30.0 = "You're fat! Lose some weight, fatty!" | otherwise   = "You're a whale, congratulations!"

In [ ]:
bmiTell 22

"You're supposedly normal. Pffft, I bet you're ugly!"

- Guards are indicated by pipes that follow a function's name and its parameters.
- A guard is basically a boolean expression. If it evaluates to True, then the corresponding function body is used. If it evaluates to False, checking drops through to the next guard and so on.

### Where
With where you can bind calculations to a name. The bmiTell function can now be written as following:
```Haskell
bmiTell :: (RealFloat a) => a -> a -> String
bmiTell weight height
| bmi <= 18.5 = "You're underweight, you emo, you!"
| bmi <= 25.0 = "You're supposedly normal. Pffft, I bet you're ugly!"
| bmi <= 30.0 = "You're fat! Lose some weight, fatty!"
| otherwise = "You're a whale, congratulations!"
where bmi = weight / height ^ 2
```
- You can bind several names.
- These names are visible across guards
- You can also use them to pattern match!
```Haskell
where bmi = weight / height ^ 2
(skinny, normal, fat) = (18.5, 25.0, 30.0)
```
another example:
```Haskell
initials :: String -> String -> String
initials firstname lastname = [f] ++ ". " ++ [l] ++ "."
    where (f:_) = firstname
          (l:_) = lastname
```

### let
- Lets you bind to variables everywhere and is an expression itself.
For example:
```Haskell
cylinder :: (RealFloat a) => a -> a -> a
cylinder r h =
    let sideArea = 2 * pi * r * h
        topArea = pi * r ^2
    in  sideArea + 2 * topArea
```
- The form is let $<bindings> in <expression>$. 

### Case (expressions)
Patter: 
```
case expression of pattern -> result
pattern -> result
pattern -> result
```
Example:
```Haskell
describeList :: [a] -> String
describeList xs = "The list is " ++ case xs of [] -> "empty."
                                               [x] -> "a singleton list."
                                               xs -> "a longer list."
```

## Recursion

## Higher order functions


### Curried functions
- Every function in Haskell only takes one parameter.
So how is it possible that we defined and used several functions that take more than one parameter so far? 
- All the functions that accepted several parameters so far have been curried functions. What does that
Example: max 4 5 
- It looks like it takes two parameters and returns the one that's bigger. 
- Max 4 5 first creates a function that takes a parameter and returns either 4 or that parameter, depending on which is bigger. 
- Then, 5 is applied to that function and that function produces our desired result.
The following two calls are equivalent:

In [1]:
max 4 5

5

In [2]:
(max 4) 5

Line 1: Redundant bracket
Found:
(max 4) 5
Why not:
max 4 5

5

Lets look at the type of max

In [4]:
:t max

max :: forall a. Ord a => a -> a -> a

This can also be written as:
```Haskell
max : : (Ord a) => a -> (a -> a) 
```
So: max takes an _a_ and returns (that's the ->) a function that takes an
_a_ and returns an _a_. That's why the return type and the parameters of
functions are all simply separated with arrows.
- This enables to partially apply a function.
Example:

In [5]:
multThree :: (Num a) => a -> a -> a -> a
multThree x y z = x * y * z

In [7]:
multThree 3 5 9

135

What happens?
1. 3 is applied to multThree.
2. That creates a function that takes one parameter and returns a function.
3. 5 is applied to that. -> creates a function that will take a parameter and multiply it by 15
4. 9 is applied to that function and the result is 135 or something.

Remember that this function's type could also be written as:
```Haskell
multThree : : (Num a) => a -> (a ->(a -> a) )
```
- The thing before the -> is the parameter that a function takes and the thing after it is what it returns. 
- So our function takes an a and returns a function of type: 
```Haskell
(Num a) => a -> (a -> a). 
```
- This function takes an a and returns a function of type:
```Haskell
(Num a) => a -> a
```
- And this function, finally, just takes an a and returns an a. Take a look at this:

You can also shorten function definitions:

In [8]:
compareWithHundred :: (Num a, Ord a) => a -> Ordering
compareWithHundred x = compare 100 x

If varables are on both sides of the equation you can shorten them:

In [9]:
compareWithHundred :: (Num a, Ord a) => a -> Ordering
compareWithHundred = compare 100

In [10]:
compareWithHundred 99

GT

This is possible becuase compareWithHundred returns a functions takes a number and compares it with 100.
Another example with and infix function:

In [13]:
devideByTen :: (Floating a) => a -> a
devideByTen = (/10)

In [14]:
devideByTen 100

10.0

- This creates a function that takes one parameter and then applies it to the side that's missing an operand.